File: CSES-sheet/Graph/Monsters-task-1194/code.java
//https://cses.fi/problemset/task/1194/
import java.io.*;
import java.util.*;

public class Demo {

  
    static char[][] adj;
    static List<Integer> cycle;
    static char[] dir ={'U','R','D','L'};
    static int[] drow={-1,0,1,0};
    static int[] dcol={0,1,0,-1};


    public static boolean isBoundary(int row,int col,int n,int m){
        return (row==n-1 || row==0 || col==0 || col==m-1);
    }

    public static boolean isValid(int r,int c,int n,int m){
        return (r>=0 && r<n && c>=0 && c<m);
    }

    
    
  

    public static void solve(Scanner sc) {
      

        int n = sc.nextInt();
        int m = sc.nextInt();

          adj = new char[n][m];

    

        for (int i = 0; i < n; i++) {
            String line=sc.next();
           
          for(int j=0;j<line.length();j++){
            adj[i][j]=line.charAt(j);
          }
         
        }

        boolean[][] vis =new boolean[n][m];

    
       Queue<int[]> MonsterQueue =new LinkedList<>();

        int[][] dist_monster=new int[n][m];

    int INF=Integer.MAX_VALUE;

    for(int[] r:dist_monster){
        Arrays.fill(r, INF);
    }

       for(int i=0;i<n;i++){
         for(int j=0;j<m;j++){
            if(adj[i][j]=='M'){
                MonsterQueue.add(new int[]{i,j});
                vis[i][j]=true;
                dist_monster[i][j]=0;
         }else if(adj[i][j]=='#'){
            vis[i][j]=true;
         }
       } 
    }



   


    while(!MonsterQueue.isEmpty()){
        int size=MonsterQueue.size();

        while (size-->0) {
            int[] curr=MonsterQueue.poll();

            int row=curr[0];
            int col=curr[1];

            for(int k=0;k<4;k++){
               int nr=row+drow[k];
               int nc=col+dcol[k];
                if(isValid(nr,nc,n,m) && !vis[nr][nc]){
                    vis[nr][nc]=true;
                    MonsterQueue.add(new int[]{nr,nc});
                    dist_monster[nr][nc]=dist_monster[row][col]+1;
                }
            }
        }
    }


     Queue<int[]> Queue =new LinkedList<>();

        int[][] dist=new int[n][m];

    char[][] parent=new char[n][m];

     boolean[][] CharVis =new boolean[n][m];
      int[] st={-1,-1};

  

       for(int i=0;i<n;i++){
         for(int j=0;j<m;j++){
            if(adj[i][j]=='A'){
                Queue.add(new int[]{i,j});
                CharVis[i][j]=true;
                st[0]=i;
                st[1]=j;
                dist[i][j]=0;
         }else if(adj[i][j]=='#'){
            CharVis[i][j]=true;
         }
       } 
    }

   
   int[] end={-1,-1};
   

    boolean isPossible=false;
    while(!Queue.isEmpty()){
        int size=Queue.size();

        while (size-->0) {
            int[] curr=Queue.poll();

            int row=curr[0];
            int col=curr[1];

            if(isBoundary(row,col,n,m)){
                end[0]=row;
                end[1]=col;
                isPossible=true;
                break;
            }

            for(int k=0;k<4;k++){
               int nr=row+drow[k];
               int nc=col+dcol[k];
                if(isValid(nr,nc,n,m) && !CharVis[nr][nc] && dist[row][col]+1<dist_monster[nr][nc]){
                    parent[nr][nc]=dir[k];
                    CharVis[nr][nc]=true;
                    Queue.add(new int[]{nr,nc});
                    dist[nr][nc]=dist[row][col]+1;
                }
            }
        }
    }



    if(!isPossible){
        System.out.println("NO");
        return;
    }

    
    int row=end[0];
    int col=end[1];

    StringBuilder sb=new StringBuilder("");

    while(!(row==st[0] && col==st[1])){
        char c =parent[row][col];
       
       sb.append(c);

        if(c=='U')row++;
        else if(c=='R')col--;
        else if(c=='D')row--;
        else if(c=='L')col++;
    }

    System.out.println("YES");
    System.out.println(sb.length());
    System.out.println(sb.reverse().toString().trim());
    

  





    





 


   
        
    }

    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        solve(sc);
        sc.close();
    }
}

File: CSES-sheet/Graph/Monsters-task-1194/code.cpp
#include <bits/stdc++.h>
using namespace std;

const int INF = 1e9;

int n, m;
vector<string> grid;

int dr[] = {-1, 0, 1, 0};
int dc[] = {0, 1, 0, -1};
char dir[] = {'U', 'R', 'D', 'L'};

struct Point {
    int r, c;
};

bool isValid(int r, int c) {
    return r >= 0 && r < n && c >= 0 && c < m;
}

bool isBoundary(int r, int c) {
    return r == 0 || c == 0 || r == n - 1 || c == m - 1;
}

vector<vector<int>> bfsMonsters() {

    vector<vector<int>> monsterDist(n, vector<int>(m, INF));
    vector<vector<bool>> vis(n, vector<bool>(m, false));

    queue<Point> q;

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {

            if (grid[i][j] == 'M') {
                q.push({i, j});
                vis[i][j] = true;
                monsterDist[i][j] = 0;
            }
            else if (grid[i][j] == '#') {
                vis[i][j] = true;
            }
        }
    }

    while (!q.empty()) {

        Point cur = q.front();
        q.pop();

        for (int k = 0; k < 4; k++) {

            int nr = cur.r + dr[k];
            int nc = cur.c + dc[k];

            if (!isValid(nr, nc) || vis[nr][nc])
                continue;

            vis[nr][nc] = true;
            monsterDist[nr][nc] = monsterDist[cur.r][cur.c] + 1;
            q.push({nr, nc});
        }
    }

    return monsterDist;
}

void bfsPlayer(Point start, vector<vector<int>>& monsterDist) {

    queue<Point> q;

    vector<vector<bool>> vis(n, vector<bool>(m, false));
    vector<vector<int>> dist(n, vector<int>(m, 0));
    vector<vector<char>> parent(n, vector<char>(m));

    q.push(start);
    vis[start.r][start.c] = true;

    Point end = {-1, -1};

    while (!q.empty()) {

        Point cur = q.front();
        q.pop();

        if (isBoundary(cur.r, cur.c)) {
            end = cur;
            break;
        }

        for (int k = 0; k < 4; k++) {

            int nr = cur.r + dr[k];
            int nc = cur.c + dc[k];

            if (!isValid(nr, nc))
                continue;

            if (vis[nr][nc] || grid[nr][nc] == '#')
                continue;

            if (dist[cur.r][cur.c] + 1 >= monsterDist[nr][nc])
                continue;

            vis[nr][nc] = true;
            dist[nr][nc] = dist[cur.r][cur.c] + 1;
            parent[nr][nc] = dir[k];

            q.push({nr, nc});
        }
    }

    if (end.r == -1) {
        cout << "NO\n";
        return;
    }

    string path;

    int r = end.r;
    int c = end.c;

    while (!(r == start.r && c == start.c)) {

        char ch = parent[r][c];
        path += ch;

        if (ch == 'U')
            r++;
        else if (ch == 'D')
            r--;
        else if (ch == 'L')
            c++;
        else
            c--;
    }

    reverse(path.begin(), path.end());

    cout << "YES\n";
    cout << path.size() << '\n';
    cout << path << '\n';
}

int main() {

    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n >> m;

    grid.resize(n);

    Point start;

    for (int i = 0; i < n; i++) {
        cin >> grid[i];

        for (int j = 0; j < m; j++) {
            if (grid[i][j] == 'A')
                start = {i, j};
        }
    }

    vector<vector<int>> monsterDist = bfsMonsters();

    bfsPlayer(start, monsterDist);

    return 0;
}
